beep/protocol/biologic_mb_to_maccor.py

```diff
@@ -398,3 +500,247 @@ def _create_step(
         )
 
         return step
+
+    @classmethod
+    def _create_steps(cls, seqs):
+        # DANGER
+        #
+        # This code has a lot of complexity and subtlties I couldn't figure out
+        # how to remove
+        #
+        # all of the various passes over the list of sequences are _very_ tightly coupled
+
+        # start build building up a mapping of where loops start and stop
+        # since loops are defined at their end instead of beginning, we will
+        # iterate from the right.
+        #
+        # during this process we want to assert loop overlap invariants
+
+        containing_loop_range_by_seq_num = {}
+        seq_num_is_non_empty_loop_start = set()
+        curr_loop_start = float("inf")
+        curr_loop_end = float("inf")
+
+        seqs_in_reverse_order = seqs.copy()
+        list.reverse(seqs_in_reverse_order)
+        for seq in seqs_in_reverse_order:
+            is_loop_seq = seq["ctrl_type"] == "Loop"
+            seq_num = int(seq["Ns"])
+            loop_to = int(seq["ctrl_seq"])
+            loop_is_empty = seq["ctrl_repeat"] == "0"
```

Comment:
I seem to remember that if the repeat is 0 the cycler will just execute the next step, but it will also not increment the loop counter (so that the cycle index won't increment), is that right?

Reply:
That is correct, if a loop has the repeat set to zero it will be immediately passed over with no effect to cycle number. The only use for such a seq, as far as I'm aware, is in the construction of an "Advance Cycle" step.

Reply:
Yeah, although there need to be two loop steps to create the equivalent of the "AdvCycle" step, right?

Reply:
That is correct, this code allows for that specifically.  As an example of adding an "Advance Cycle" step in the EC-Lab software:
![image](https://user-images.githubusercontent.com/5530247/94617383-4ef82a00-025e-11eb-970f-ed94d98763da.png)
![image](https://user-images.githubusercontent.com/5530247/94617398-561f3800-025e-11eb-9600-5a2d5fd42c7f.png)
This adds a single advance cycle between seqs 0 and 3.
 
I've pushed a small naming change to alter this from "empty_loop" to "inactive_loop" and "non_empty_loop" to "active_loop", which I hope should be clearer.